File: backend/aitrade/models/governance.py
"""
CNN 模型治理请求/响应模型。

定义 Walk-Forward 评估、候选训练、晋级/回滚、治理回放等治理流程的 Pydantic v2 模型。
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

from .alpha import LabelSpec, ObservationGroup


GovernanceBaseline = Literal[
    "fixed_initial_model",
    "always_retrain",
    "governed_promotion",
    "buy_and_hold",
]


class CNNGovernanceConfig(BaseModel):
    """CNN 治理模块配置（全局开关与默认参数）。

    控制 Walk-Forward 评估周期、训练/测试窗口大小、随机种子数量等治理行为；
    ``auto_promote=False`` 要求人工确认晋级，适合第一版保守运营。
    """

    enabled: bool = Field(default=True, description="是否启用治理模块")
    evaluation_period_days: int = Field(default=30, ge=1, description="默认评估周期（天）")
    train_days: int = Field(default=720, ge=30, description="默认训练窗口（天）")
    test_days: int = Field(default=90, ge=1, description="默认 OOS 测试窗口（天）")
    n_seeds: int = Field(default=1, ge=1, le=10, description="默认随机种子数量")
    next_suggested_eval_date: Optional[date] = Field(default=None, description="下一次建议评估日期")
    auto_promote: bool = Field(default=False, description="是否自动晋级；第一版固定建议 false")


class CNNTrainingParams(BaseModel):
    """CNN 治理训练超参数（Walk-Forward 与候选训练共用）。

    与单次训练接口的参数结构对齐；各字段均带范围约束，防止异常值导致 OOM 或发散。
    """

    epochs: int = Field(default=20, ge=1, le=500)
    batch_size: int = Field(default=32, ge=1, le=4096)
    learning_rate: float = Field(default=0.001, gt=0)
    lookback: int = Field(default=30, ge=2, le=4096)
    dropout: float = Field(default=0.4, ge=0, lt=1)
    train_ratio: float = Field(default=0.8, gt=0, lt=1)
    # 损失加权策略：none=普通 BCE，各样本权重相同；
    # magnitude=按 |未来收益| 对样本加权，让大波动样本主导梯度（仅分类目标生效）。
    loss_weighting: Literal["none", "magnitude"] = "none"


class CNNBacktestParams(BaseModel):
    """CNN 治理回测参数（阈值、成本、出场规则）。

    控制信号阈值（buy_threshold / sell_threshold）、A 股交易成本（佣金/印花税/滑点）
    以及出场模式（threshold / fixed_hold / oco / auto）；T+1 默认开启贴近 A 股现实。

    ``veto_threshold`` 仅在 objective='path_class' 时生效：当「先触止损」类概率
    prob_sl >= 该值时否决买入；默认 1.0 等效关闭（向后兼容）。
    """

    buy_threshold: float = 0.6
    sell_threshold: float = 0.4
    commission_rate: float = Field(default=0.0003, ge=0, lt=0.1)
    stamp_duty: float = Field(default=0.0005, ge=0, lt=0.1)
    slippage: float = Field(default=0.0005, ge=0, lt=0.1)
    price_add: float = Field(default=0.002, ge=0, lt=0.1)
    exit_mode: Literal["threshold", "fixed_hold", "oco", "auto"] = "auto"
    hold_days: int = Field(default=1, ge=1, le=60)
    take_profit: float = Field(default=0.0, ge=0, lt=1)
    stop_loss: float = Field(default=0.0, ge=0, lt=1)
    t_plus1: bool = False
    veto_threshold: float = Field(
        default=1.0,
        gt=0.0,
        le=1.0,
        description=(
            "path_class 专用：先触止损概率 prob_sl >= 该值时否决买入；"
            "默认 1.0 等效关闭否决（向后兼容）。"
        ),
    )


class CNNPromotionGate(BaseModel):
    """候选模型晋级门禁（Walk-Forward 折内胜出条件）。

    当候选模型的 min_win_rate 折数胜出、核心指标提升 >= min_core_score_delta
    且最大回撤劣化 <= max_drawdown_worsen_pct 时，才允许晋级（或提示运营人工确认）。
    """

    min_win_rate: float = Field(default=0.5, ge=0, le=1, description="候选胜出折数比例下限")
    min_core_score_delta: float = Field(default=0.0, description="核心分数相对生产模型的平均提升下限")
    max_drawdown_worsen_pct: float = Field(default=10.0, ge=0, description="最大回撤允许劣化百分比")
    require_positive_oos: bool = Field(default=True, description="是否要求 OOS 核心指标为正")


class CNNWalkForwardRequest(BaseModel):
    """启动 CNN Walk-Forward / OOS 评估的请求体。

    分 train_days 训练窗 + test_days OOS 测试窗逐步前推；production_model 为空时
    自动读取当前生产模型作为相对胜出基准。step_days 为空时默认等于 test_days（无重叠滚动）。
    """

    name: str = Field(description="评估名称")
    target_symbol: str
    input_data_kind: str = "bar"
    input_interval: str = "d"
    start: date
    end: date
    train_days: int = Field(default=720, ge=30)
    test_days: int = Field(default=90, ge=1)
    step_days: Optional[int] = Field(default=None, ge=1)
    n_seeds: int = Field(
        default=1,
        ge=1,
        le=10,
        description="每折重复训练的随机种子数；1=单种子，>1 时折内对多个种子取均值并衡量跨种子波动",
    )
    objective: Literal["classification", "regression", "path_class"] = Field(
        default="classification",
        description=(
            "预测目标：classification=方向二分类；regression=涨跌幅回归；"
            "path_class=路径形态四分类（先触止盈/先触止损/到期小涨/到期小跌，需配合 label_spec.mode='oco'）"
        ),
    )
    label_spec: LabelSpec = Field(default_factory=LabelSpec)
    observation_groups: list[ObservationGroup] = Field(default_factory=list)
    training_params: CNNTrainingParams = Field(default_factory=CNNTrainingParams)
    backtest_params: CNNBacktestParams = Field(default_factory=CNNBacktestParams)
    promotion_gate: CNNPromotionGate = Field(default_factory=CNNPromotionGate)
    production_model: Optional[str] = Field(default=None, description="用于相对胜出的生产模型；为空则读取当前生产模型")


class CNNCandidateTrainRequest(CNNWalkForwardRequest):
    """在全量（或指定）窗口上训练候选模型的请求体。

    继承 ``CNNWalkForwardRequest`` 的所有评估配置；额外可指定 final_train_start /
    final_train_end 以覆盖最终训练窗，不指定时沿用请求 start 或最后一个训练窗结束。
    """

    final_train_start: Optional[date] = Field(default=None, description="最终候选训练起始日期；为空用请求 start")
    final_train_end: Optional[date] = Field(default=None, description="最终候选训练结束日期；为空用最后一个训练窗口结束")


class CNNPromotionRequest(BaseModel):
    """将候选模型晋级为生产模型的请求体（人工确认用）。"""

    promoted_by: str = Field(default="manual")
    note: str = ""


class CNNRollbackRequest(BaseModel):
    """将生产模型回滚到指定版本（或上一生产模型）的请求体。"""

    rollback_to: Optional[str] = Field(default=None, description="指定回滚模型；为空回滚上一生产模型")
    requested_by: str = Field(default="manual")
    note: str = ""


class CNNGovernanceReplayRequest(BaseModel):
    """治理回放回测请求：模拟历史上若使用治理流程会有何收益/风险。

    以 initial_train_days 初始训练窗启动，每 evaluation_period_days 评估一次，
    并与 baselines（固定初始模型/永远重训/治理晋级/买入持有）做横向对比。
    """

    name: str
    target_symbol: str
    input_data_kind: str = "bar"
    input_interval: str = "d"
    start: date
    end: date
    initial_train_days: int = Field(default=720, ge=30)
    evaluation_period_days: int = Field(default=30, ge=1)
    test_period_days: int = Field(default=30, ge=1)
    capital: float = Field(default=1_000_000, gt=0)
    objective: Literal["classification", "regression", "path_class"] = Field(
        default="classification",
        description=(
            "预测目标：classification=方向二分类；regression=涨跌幅回归；"
            "path_class=路径形态四分类（先触止盈/先触止损/到期小涨/到期小跌，需配合 label_spec.mode='oco'）"
        ),
    )
    label_spec: LabelSpec = Field(default_factory=LabelSpec)
    observation_groups: list[ObservationGroup] = Field(default_factory=list)
    training_params: CNNTrainingParams = Field(default_factory=CNNTrainingParams)
    backtest_params: CNNBacktestParams = Field(default_factory=CNNBacktestParams)
    promotion_gate: CNNPromotionGate = Field(default_factory=CNNPromotionGate)
    baselines: list[GovernanceBaseline] = Field(
        default_factory=lambda: [
            "fixed_initial_model",
            "always_retrain",
            "governed_promotion",
            "buy_and_hold",
        ]
    )


class CNNProductionModel(BaseModel):
    """当前生产模型状态（持久化的晋级记录快照）。

    promoted_at / promoted_by 记录最近一次晋级操作的时刻与操作者；
    previous_model_name / previous_model_version 保留上一版生产模型信息（回滚用）。
    """

    model_name: str = ""
    model_version: str = ""
    target_symbol: str = ""
    input_interval: str = "d"
    objective: str = "classification"
    promoted_at: Optional[datetime] = None
    promoted_by: str = ""
    report_id: str = ""
    previous_model_name: str = ""
    previous_model_version: str = ""


class CNNGovernanceHistoryEvent(BaseModel):
    """治理历史事件记录（审计日志条目）。

    每次晋级/回滚/评估触发时追加到历史序列；payload 含操作参数与判定结果，
    供治理回放与合规审计使用。
    """

    ts: datetime = Field(default_factory=datetime.now)
    event_type: str
    payload: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/types/governance.ts
import type { LabelSpec, ObservationGroup, TaskStartResponse } from './alpha'

/**
 * 治理回演基准类型：与当前候选模型比较的基准策略。
 *
 * - `fixed_initial_model`：固定初始模型不更新；
 * - `always_retrain`：每期强制重训；
 * - `governed_promotion`：治理晋升策略；
 * - `buy_and_hold`：买入持有基准。
 */
export type GovernanceBaseline =
  | 'fixed_initial_model'
  | 'always_retrain'
  | 'governed_promotion'
  | 'buy_and_hold'

/** CNN 治理全局配置（`GET /PUT /api/cnn/governance/config`），控制评估周期与默认窗口。 */
export interface CNNGovernanceConfig {
  /** 是否启用治理模块；关闭后不再触发自动评估/晋级 */
  enabled: boolean
  /** 默认评估周期，单位天 */
  evaluation_period_days: number
  /** 默认训练窗口，单位天 */
  train_days: number
  /** 默认 OOS（样本外）测试窗口，单位天 */
  test_days: number
  /** 默认每折重复训练的随机种子数；1=单种子，>1 时折内对多种子取均值并衡量波动 */
  n_seeds: number
  /** 系统建议的下一次评估日期（YYYY-MM-DD）；无建议时为 null */
  next_suggested_eval_date?: string | null
  /** 是否自动晋级；false 时需人工确认，第一版保守运营默认 false */
  auto_promote: boolean
}

/** CNN 训练超参数集合（用于治理/候选训练请求），与单次训练接口的参数结构对齐。 */
export interface CNNTrainingParams {
  /** 训练轮数 */
  epochs: number
  /** 批大小 */
  batch_size: number
  /** 学习率 */
  learning_rate: number
  /** 输入序列回看长度（时间步数） */
  lookback: number
  /** Dropout 比例，取值 [0, 1) */
  dropout: number
  /** 训练集占比，取值 (0, 1)，其余为验证集 */
  train_ratio: number
  /** 损失加权：none=不加权；magnitude=按涨跌幅度加权 */
  loss_weighting: 'none' | 'magnitude'
}

/** CNN 回测参数集合（用于治理/候选评估请求）：阈值、A 股交易成本与出场规则。 */
export interface CNNBacktestParams {
  /** 买入信号概率阈值 */
  buy_threshold: number
  /** 卖出信号概率阈值 */
  sell_threshold: number
  /** 单边佣金率（如 0.0003=万3） */
  commission_rate: number
  /** 卖出印花税率（如 0.0005=千0.5，A 股） */
  stamp_duty: number
  /** 每笔成交不利滑点率 */
  slippage: number
  /** 限价单价格缓冲/市价化挂单比例 */
  price_add: number
  /** 出场模式：threshold=概率阈值；fixed_hold=固定持有；oco=止盈止损；auto=按 label 自动对齐 */
  exit_mode: 'threshold' | 'fixed_hold' | 'oco' | 'auto'
  /** fixed_hold/oco 的固定/最大持有交易日数 */
  hold_days: number
  /** oco 止盈幅度（0.02=+2%），0=不启用 */
  take_profit: number
  /** oco 止损幅度（0.03=-3%），0=不启用 */
  stop_loss: number
  /** 是否启用 T+1 卖出限制（A 股现实约束） */
  t_plus1: boolean
}

/**
 * CNN 候选模型晋升门槛配置（Walk-Forward 折内胜出条件）。
 *
 * 候选需同时满足：胜出折数比例 >= min_win_rate、核心分数提升 >= min_core_score_delta、
 * 最大回撤劣化 <= max_drawdown_worsen_pct，方可晋级（或提示人工确认）。
 */
export interface CNNPromotionGate {
  /** 候选胜出折数比例下限，取值 [0, 1] */
  min_win_rate: number
  /** 核心分数相对生产模型的平均提升下限（绝对值，可为负以放宽） */
  min_core_score_delta: number
  /** 最大回撤允许劣化的百分比上限 */
  max_drawdown_worsen_pct: number
  /** 是否要求 OOS（样本外）核心指标为正 */
  require_positive_oos: boolean
}

/**
 * CNN Walk-Forward 评估请求（`POST /api/cnn/governance/evaluate`）。
 *
 * 以 train_days 训练窗 + test_days OOS 测试窗逐步前推，多折评估候选相对生产模型的胜出情况。
 */
export interface CNNWalkForwardRequest {
  /** 评估任务名称 */
  name: string
  /** 目标合约代码 */
  target_symbol: string
  /** 输入数据类型：bar=K 线；tick=逐笔 */
  input_data_kind: 'bar' | 'tick'
  /** 输入数据周期，如 "d"（日线）、"1m"/"30m"（分钟线） */
  input_interval: string
  /** 评估区间起始日期（YYYY-MM-DD，含） */
  start: string
  /** 评估区间结束日期（YYYY-MM-DD，含） */
  end: string
  /** 单折训练窗口长度（天） */
  train_days: number
  /** 单折 OOS 测试窗口长度（天） */
  test_days: number
  /** 相邻折前推步长（天）；为空时默认等于 test_days（无重叠滚动） */
  step_days?: number
  /** 预测目标：classification=方向二分类；regression=涨跌幅回归 */
  objective: 'classification' | 'regression'
  /** 标签生成规格（涨跌阈值/持有窗等） */
  label_spec: LabelSpec
  /** 观测特征分组配置 */
  observation_groups: ObservationGroup[]
  /** 训练超参数 */
  training_params: CNNTrainingParams
  /** 回测参数（阈值/成本/出场） */
  backtest_params: CNNBacktestParams
  /** 候选晋级门槛 */
  promotion_gate: CNNPromotionGate
  /** 用于相对胜出比较的生产模型；为空则读取当前生产模型 */
  production_model?: string
}

/** CNN 候选模型完整训练请求（`POST /api/cnn/governance/candidates/train`），在 Walk-Forward 基础上增加最终训练区间。 */
export interface CNNCandidateTrainRequest extends CNNWalkForwardRequest {
  /** 最终候选训练起始日期（YYYY-MM-DD）；为空时沿用请求 start */
  final_train_start?: string
  /** 最终候选训练结束日期（YYYY-MM-DD）；为空时沿用最后一个训练窗口结束 */
  final_train_end?: string
}

/**
 * CNN 治理历史重演请求（`POST /api/cnn/governance/replay/run`）。
 *
 * 模拟历史上若启用治理流程的收益/风险：以 initial_train_days 启动，每 evaluation_period_days
 * 评估一次，并与 baselines 中的各基准策略横向对比。
 */
export interface CNNGovernanceReplayRequest {
  /** 重演任务名称 */
  name: string
  /** 目标合约代码 */
  target_symbol: string
  /** 输入数据类型：bar=K 线；tick=逐笔 */
  input_data_kind: 'bar' | 'tick'
  /** 输入数据周期，如 "d"、"30m" */
  input_interval: string
  /** 重演区间起始日期（YYYY-MM-DD，含） */
  start: string
  /** 重演区间结束日期（YYYY-MM-DD，含） */
  end: string
  /** 启动时的初始训练窗口长度（天） */
  initial_train_days: number
  /** 每隔多少天评估并考虑晋级一次（天） */
  evaluation_period_days: number
  /** 每次评估的 OOS 测试窗口长度（天） */
  test_period_days: number
  /** 初始资金，单位元 */
  capital: number
  /** 预测目标：classification=方向二分类；regression=涨跌幅回归 */
  objective: 'classification' | 'regression'
  /** 标签生成规格 */
  label_spec: LabelSpec
  /** 观测特征分组配置 */
  observation_groups: ObservationGroup[]
  /** 训练超参数 */
  training_params: CNNTrainingParams
  /** 回测参数 */
  backtest_params: CNNBacktestParams
  /** 候选晋级门槛 */
  promotion_gate: CNNPromotionGate
  /** 参与横向对比的基准策略集合 */
  baselines: GovernanceBaseline[]
}

/**
 * 当前生产模型状态（`GET /api/cnn/governance/production` 响应体），即持久化的晋级记录快照。
 *
 * previous_* 字段保留上一版生产模型信息，供回滚使用。
 */
export interface CNNProductionModel {
  /** 当前生产模型名称 */
  model_name: string
  /** 当前生产模型版本号 */
  model_version: string
  /** 目标合约代码 */
  target_symbol: string
  /** 输入数据周期，如 "d" */
  input_interval: string
  /** 预测目标（classification / regression / path_class） */
  objective: string
  /** 最近一次晋级时刻（ISO 时间串）；从未晋级时为空 */
  promoted_at?: string | null
  /** 最近一次晋级的操作者 */
  promoted_by: string
  /** 晋级所依据的治理评估报告 ID */
  report_id: string
  /** 上一版生产模型名称（回滚目标） */
  previous_model_name: string
  /** 上一版生产模型版本号 */
  previous_model_version: string
}

/** CNN 候选模型记录（`GET /api/cnn/governance/candidates` 列表元素）。 */
export interface CNNCandidate {
  /** 候选记录唯一 ID */
  candidate_id: string
  /** 候选创建时刻（ISO 时间串） */
  created_at: string
  /** 候选状态：pending=待评估；passed=通过门槛；failed=未达门槛；promoted=已晋级；rejected=人工驳回 */
  status: 'pending' | 'passed' | 'failed' | 'promoted' | 'rejected'
  /** 候选模型名称 */
  model_name: string
  /** 关联的治理评估报告 ID */
  report_id: string
  /** 目标合约代码 */
  target_symbol: string
  /** 输入数据周期，如 "d" */
  input_interval: string
  /** 预测目标（classification / regression / path_class） */
  objective: string
  /** 评估时对比的基准生产模型名称；无对比时缺省 */
  baseline_model?: string
  /** 评估结果摘要（指标键值对），结构由后端决定 */
  summary?: Record<string, unknown>
  /** 触发本候选的原始训练请求快照，结构由后端决定 */
  request?: Record<string, unknown>
}

/** CNN 治理评估报告（`GET /api/cnn/governance/reports/{reportId}` 响应体）。 */
export interface CNNGovernanceReport {
  /** 报告唯一 ID */
  report_id: string
  /** 报告类型（如 walk_forward / candidate） */
  type: string
  /** 报告名称（取自评估请求的 name） */
  name: string
  /** 报告生成时刻（ISO 时间串） */
  created_at: string
  /** 评估时对比的生产模型名称；首次评估无对比时缺省 */
  production_model?: string
  /** 逐折评估明细，每个元素为一折的指标键值对，结构由后端决定 */
  folds: Array<Record<string, unknown>>
  /** 跨折汇总指标与晋级判定结果，结构由后端决定 */
  summary: Record<string, unknown>
}

/** CNN 治理历史重演结果报告（`GET /api/cnn/governance/replay/{replayId}` 响应体）。 */
export interface CNNGovernanceReplayReport {
  /** 重演任务唯一 ID */
  replay_id: string
  /** 重演任务名称 */
  name: string
  /** 目标合约代码 */
  target_symbol: string
  /** 重演区间起始日期（YYYY-MM-DD） */
  start: string
  /** 重演区间结束日期（YYYY-MM-DD） */
  end: string
  /** 报告生成时刻（ISO 时间串） */
  created_at: string
  /** 各基准策略的回测结果，键为基准名（见 GovernanceBaseline），值为指标键值对 */
  baselines: Record<string, Record<string, unknown>>
  /** 治理流程逐期执行轨迹，结构由后端决定 */
  process: Record<string, unknown>
  /** 重演期间发生的晋级事件列表，结构由后端决定 */
  promotion_events: Array<Record<string, unknown>>
  /** 重演期间被驳回（未达门槛）的候选事件列表，结构由后端决定 */
  rejected_events: Array<Record<string, unknown>>
  /** 诊断信息（数据缺口、异常折等），结构由后端决定 */
  diagnostics: Record<string, unknown>
  /** 总体结论（治理 vs 各基准的优劣判定），结构由后端决定 */
  conclusion: Record<string, unknown>
}

/** 治理历史事件条目（`GET /api/cnn/governance/history` 列表元素），即审计日志记录。 */
export interface GovernanceHistoryEvent {
  /** 事件发生时刻（ISO 时间串） */
  ts: string
  /** 事件类型，如 promote（晋级）/ rollback（回滚）/ evaluate（评估） */
  event_type: string
  /** 事件载荷，含操作参数与判定结果，结构随 event_type 而定 */
  payload: Record<string, unknown>
}

export type { TaskStartResponse }
